File: fabric/src/main/java/juuxel/adorn/platform/fabric/FabricRegistrarFactory.java
package juuxel.adorn.platform.fabric;

import juuxel.adorn.lib.registry.BlockRegistrar;
import juuxel.adorn.lib.registry.ItemRegistrar;
import juuxel.adorn.lib.registry.KeyedRegistrar;
import juuxel.adorn.lib.registry.Registrar;
import juuxel.adorn.lib.registry.RegistrarFactory;
import juuxel.adorn.lib.registry.TrackedDataHandlerRegistrar;
import net.minecraft.core.Registry;
import net.minecraft.core.registries.BuiltInRegistries;
import net.minecraft.network.syncher.EntityDataSerializer;
import net.minecraft.resources.ResourceKey;

public final class FabricRegistrarFactory implements RegistrarFactory {
    @SuppressWarnings("unchecked")
    @Override
    public <T> KeyedRegistrar<T> create(ResourceKey<Registry<T>> registryKey) {
        var registry = ((Registry<Registry<T>>) BuiltInRegistries.REGISTRY).getValue(registryKey);
        if (registry == null) throw new IllegalArgumentException("No registry found for key " + registryKey);
        return new RegistrarImpl<>(registry);
    }

    @Override
    public BlockRegistrar createBlocks() {
        return new RegistrarImpl.BlockImpl();
    }

    @Override
    public ItemRegistrar createItems() {
        return new RegistrarImpl.ItemImpl();
    }

    @Override
    public Registrar<EntityDataSerializer<?>> createForTrackedDataHandlers() {
        return new TrackedDataHandlerRegistrar();
    }
}

File: fabric/src/main/kotlin/juuxel/adorn/platform/fabric/FabricRegistrarFactory.kt
package juuxel.adorn.platform.fabric

import juuxel.adorn.lib.registry.Registrar
import juuxel.adorn.lib.registry.RegistrarFactory
import net.minecraft.registry.Registries
import net.minecraft.registry.Registry
import net.minecraft.registry.RegistryKey

class FabricRegistrarFactory : RegistrarFactory {
    @Suppress("UNCHECKED_CAST")
    override fun <T> create(registryKey: RegistryKey<Registry<T>>): Registrar<T> {
        val registry = (Registries.REGISTRIES as Registry<Registry<T>>).get(registryKey)
            ?: throw IllegalArgumentException("No registry found for key $registryKey")
        return RegistrarImpl(registry)
    }
}
